aiida/orm/implementation/sqlalchemy/utils.py

```diff
@@ -80,11 +80,14 @@ def __setattr__(self, key, value):
             self._flush(fields=fields)
 
     def is_saved(self):
-        """Retun whether the wrapped model instance is saved in the database.
+        """Return whether the wrapped model instance is saved in the database.
 
         :return: boolean, True if the model is saved in the database, False otherwise
         """
-        return self._model.id is not None
+        # we should not flush here since it may lead to IntegrityErrors
+        # which are handled later in the save method
```

Comment:
```suggestion
        # We should disable the automatic flush here, which is enabled by default in SqlAlchemy
        # and each query (including just fetching an attribute) will flush the session. However, here
        # a flush may lead to `IntegrityError`'s, which should instead be handled in the `save` method.
```

Reply:
> including just fetching an attribute

as I've mentioned above, this is not case; it is because of our overrides of the `__getattr__`, `__setattr__` methods, and the fact we are calling this fetch "within" a setattr action

Reply:
Are you sure though? Since the `id` is an immutable property, it should not result in a database hit. We have an exception for this in the `ModelWrapper`. So if what I am saying is incorrect, then why _is_ it flushing? That is the thing that I want to add here.

Reply:
Anyway, I will approve now and let you improve the message as you see fit. I am still extremely befuddled why exactly this is happening.

Reply:
> Are you sure though? Since the id is an immutable property, it should not result in a database hit.

yep pretty sure (having recently watched a 3 hour video on sqla lol, which I highly recommend https://youtu.be/1Va493SMTcY); in the failing test line `group_b.label = label_group_a`, the first thing we do (before reaching `is_saved`) is set the attribute on the model (`setattr(self._model, key, value)`). Sqlachemy has now marked the model as "dirty" and is primed to flush on the next getattr, which we then do with `self._model.id`.

Note you could add: `sessionmaker(bind=engine, autoflush=False)`, but I would be wary of changing this

Very simply, this is what is happening:

```python
In [1]: session.commit()

In [2]: session.dirty
Out[2]: IdentitySet([])

In [3]: user.name = object()

In [4]: session.dirty
Out[4]: IdentitySet([<__main__.User object at 0x7f859fd25d00>])

In [5]: session.id
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-84-b8bfe9342bd2> in <module>
----> 1 session.id

AttributeError: 'Session' object has no attribute 'id'

In [85]: user.id
2021-09-09 12:13:25,521 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-09-09 12:13:25,522 INFO sqlalchemy.engine.Engine SELECT user.id AS user_id 
FROM user 
WHERE user.id = ?
2021-09-09 12:13:25,523 INFO sqlalchemy.engine.Engine [cached since 527.9s ago] (1,)
2021-09-09 12:13:25,523 INFO sqlalchemy.engine.Engine UPDATE user SET name=? WHERE user.id = ?
2021-09-09 12:13:25,524 INFO sqlalchemy.engine.Engine [cached since 654.2s ago] (<object object at 0x7f85a2f51bf0>, 1)
2021-09-09 12:13:25,524 INFO sqlalchemy.engine.Engine ROLLBACK
---------------------------------------------------------------------------
InterfaceError                            Traceback (most recent call last)
~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1770                 if not evt_handled:
-> 1771                     self.dialect.do_execute(
   1772                         cursor, statement, parameters, context

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py in do_execute(self, cursor, statement, parameters, context)
    716     def do_execute(self, cursor, statement, parameters, context=None):
--> 717         cursor.execute(statement, parameters)
    718 

InterfaceError: Error binding parameter 0 - probably unsupported type.

The above exception was the direct cause of the following exception:

InterfaceError                            Traceback (most recent call last)
<ipython-input-85-63cd6cde9a85> in <module>
----> 1 user.id

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py in __get__(self, instance, owner)
    479                     replace_context=err,
    480                 )
--> 481             return self.impl.get(state, dict_)
    482 
    483 

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py in get(self, state, dict_, passive)
    924                     return PASSIVE_NO_RESULT
    925 
--> 926                 value = self._fire_loader_callables(state, key, passive)
    927 
    928                 if value is PASSIVE_NO_RESULT or value is NO_VALUE:

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/attributes.py in _fire_loader_callables(self, state, key, passive)
    955             and key in state.expired_attributes
    956         ):
--> 957             return state._load_expired(state, passive)
    958         elif key in state.callables:
    959             callable_ = state.callables[key]

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/state.py in _load_expired(self, state, passive)
    708         )
    709 
--> 710         self.manager.expired_attribute_loader(self, toload, passive)
    711 
    712         # if the loader failed, or this

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in load_scalar_attributes(mapper, state, attribute_names, passive)
   1443             return
   1444 
-> 1445         result = load_on_ident(
   1446             session,
   1447             future.select(mapper).set_label_style(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in load_on_ident(session, statement, key, load_options, refresh_state, with_for_update, only_load_props, no_autoflush, bind_arguments, execution_options)
    399         ident = identity_token = None
    400 
--> 401     return load_on_pk_identity(
    402         session,
    403         statement,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/loading.py in load_on_pk_identity(session, statement, primary_key_identity, load_options, refresh_state, with_for_update, only_load_props, identity_token, no_autoflush, bind_arguments, execution_options)
    522     )
    523     result = (
--> 524         session.execute(
    525             q,
    526             params=params,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in execute(self, statement, params, execution_options, bind_arguments, _parent_execute_state, _add_event, **kw)
   1635                 statement,
   1636                 execution_options,
-> 1637             ) = compile_state_cls.orm_pre_session_exec(
   1638                 self,
   1639                 statement,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/context.py in orm_pre_session_exec(cls, session, statement, params, execution_options, bind_arguments, is_reentrant_invoke)
    290 
    291         if load_options._autoflush:
--> 292             session._autoflush()
    293 
    294         return statement, execution_options

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _autoflush(self)
   2202                     "flush is occurring prematurely"
   2203                 )
-> 2204                 util.raise_(e, with_traceback=sys.exc_info()[2])
   2205 
   2206     def refresh(self, instance, attribute_names=None, with_for_update=None):

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    205 
    206         try:
--> 207             raise exception
    208         finally:
    209             # credit to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _autoflush(self)
   2191         if self.autoflush and not self._flushing:
   2192             try:
-> 2193                 self.flush()
   2194             except sa_exc.StatementError as e:
   2195                 # note we are reraising StatementError as opposed to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in flush(self, objects)
   3296         try:
   3297             self._flushing = True
-> 3298             self._flush(objects)
   3299         finally:
   3300             self._flushing = False

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _flush(self, objects)
   3436         except:
   3437             with util.safe_reraise():
-> 3438                 transaction.rollback(_capture_exception=True)
   3439 
   3440     def bulk_save_objects(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py in __exit__(self, type_, value, traceback)
     68             self._exc_info = None  # remove potential circular references
     69             if not self.warn_only:
---> 70                 compat.raise_(
     71                     exc_value,
     72                     with_traceback=exc_tb,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    205 
    206         try:
--> 207             raise exception
    208         finally:
    209             # credit to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _flush(self, objects)
   3396             self._warn_on_events = True
   3397             try:
-> 3398                 flush_context.execute()
   3399             finally:
   3400                 self._warn_on_events = False

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py in execute(self)
    454         else:
    455             for rec in topological.sort(self.dependencies, postsort_actions):
--> 456                 rec.execute(self)
    457 
    458     def finalize_flush_changes(self):

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py in execute(self, uow)
    628     @util.preload_module("sqlalchemy.orm.persistence")
    629     def execute(self, uow):
--> 630         util.preloaded.orm_persistence.save_obj(
    631             self.mapper,
    632             uow.states_for_mapper_hierarchy(self.mapper, False, False),

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py in save_obj(base_mapper, states, uowtransaction, single)
    232         )
    233 
--> 234         _emit_update_statements(
    235             base_mapper,
    236             uowtransaction,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py in _emit_update_statements(base_mapper, uowtransaction, mapper, table, update, bookkeeping)
    996                 )
    997 
--> 998                 c = connection._execute_20(
    999                     statement, multiparams, execution_options=execution_options
   1000                 )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_20(self, statement, parameters, execution_options)
   1581             )
   1582         else:
-> 1583             return meth(self, args_10style, kwargs_10style, execution_options)
   1584 
   1585     def exec_driver_sql(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/sql/elements.py in _execute_on_connection(self, connection, multiparams, params, execution_options, _force)
    321     ):
    322         if _force or self.supports_execution:
--> 323             return connection._execute_clauseelement(
    324                 self, multiparams, params, execution_options
    325             )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_clauseelement(self, elem, multiparams, params, execution_options)
   1450             linting=self.dialect.compiler_linting | compiler.WARN_LINTING,
   1451         )
-> 1452         ret = self._execute_context(
   1453             dialect,
   1454             dialect.execution_ctx_cls._init_compiled,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1812 
   1813         except BaseException as e:
-> 1814             self._handle_dbapi_exception(
   1815                 e, statement, parameters, cursor, context
   1816             )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   1993                 util.raise_(newraise, with_traceback=exc_info[2], from_=e)
   1994             elif should_wrap:
-> 1995                 util.raise_(
   1996                     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
   1997                 )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    205 
    206         try:
--> 207             raise exception
    208         finally:
    209             # credit to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1769                             break
   1770                 if not evt_handled:
-> 1771                     self.dialect.do_execute(
   1772                         cursor, statement, parameters, context
   1773                     )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py in do_execute(self, cursor, statement, parameters, context)
    715 
    716     def do_execute(self, cursor, statement, parameters, context=None):
--> 717         cursor.execute(statement, parameters)
    718 
    719     def do_execute_no_params(self, cursor, statement, context=None):

InterfaceError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.InterfaceError) Error binding parameter 0 - probably unsupported type.
[SQL: UPDATE user SET name=? WHERE user.id = ?]
[parameters: (<object object at 0x7f85a2f51bf0>, 1)]
(Background on this error at: https://sqlalche.me/e/14/rvf5)
```

whereas we want:

```python
In [87]: session.commit()

In [88]: session.dirty
Out[88]: IdentitySet([])

In [89]: user.name = object()

In [90]: session.dirty
Out[90]: IdentitySet([<__main__.User object at 0x7f859fd25d00>])

In [91]: with session.no_autoflush:
    ...:     print(user.id)
    ...: 
2021-09-09 12:17:07,045 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-09-09 12:17:07,045 INFO sqlalchemy.engine.Engine SELECT user.id AS user_id 
FROM user 
WHERE user.id = ?
2021-09-09 12:17:07,045 INFO sqlalchemy.engine.Engine [cached since 749.4s ago] (1,)
1

In [92]: session.commit()
2021-09-09 12:17:13,939 INFO sqlalchemy.engine.Engine UPDATE user SET name=? WHERE user.id = ?
2021-09-09 12:17:13,939 INFO sqlalchemy.engine.Engine [cached since 882.6s ago] (<object object at 0x7f85a2f51b00>, 1)
2021-09-09 12:17:13,940 INFO sqlalchemy.engine.Engine ROLLBACK
---------------------------------------------------------------------------
InterfaceError                            Traceback (most recent call last)
~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1770                 if not evt_handled:
-> 1771                     self.dialect.do_execute(
   1772                         cursor, statement, parameters, context

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py in do_execute(self, cursor, statement, parameters, context)
    716     def do_execute(self, cursor, statement, parameters, context=None):
--> 717         cursor.execute(statement, parameters)
    718 

InterfaceError: Error binding parameter 0 - probably unsupported type.

The above exception was the direct cause of the following exception:

InterfaceError                            Traceback (most recent call last)
<ipython-input-92-83a1dbdbd92a> in <module>
----> 1 session.commit()

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in commit(self)
   1426                 raise sa_exc.InvalidRequestError("No transaction is begun.")
   1427 
-> 1428         self._transaction.commit(_to_root=self.future)
   1429 
   1430     def prepare(self):

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in commit(self, _to_root)
    827         self._assert_active(prepared_ok=True)
    828         if self._state is not PREPARED:
--> 829             self._prepare_impl()
    830 
    831         if self._parent is None or self.nested:

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _prepare_impl(self)
    806                 if self.session._is_clean():
    807                     break
--> 808                 self.session.flush()
    809             else:
    810                 raise exc.FlushError(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in flush(self, objects)
   3296         try:
   3297             self._flushing = True
-> 3298             self._flush(objects)
   3299         finally:
   3300             self._flushing = False

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _flush(self, objects)
   3436         except:
   3437             with util.safe_reraise():
-> 3438                 transaction.rollback(_capture_exception=True)
   3439 
   3440     def bulk_save_objects(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py in __exit__(self, type_, value, traceback)
     68             self._exc_info = None  # remove potential circular references
     69             if not self.warn_only:
---> 70                 compat.raise_(
     71                     exc_value,
     72                     with_traceback=exc_tb,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    205 
    206         try:
--> 207             raise exception
    208         finally:
    209             # credit to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/session.py in _flush(self, objects)
   3396             self._warn_on_events = True
   3397             try:
-> 3398                 flush_context.execute()
   3399             finally:
   3400                 self._warn_on_events = False

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py in execute(self)
    454         else:
    455             for rec in topological.sort(self.dependencies, postsort_actions):
--> 456                 rec.execute(self)
    457 
    458     def finalize_flush_changes(self):

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py in execute(self, uow)
    628     @util.preload_module("sqlalchemy.orm.persistence")
    629     def execute(self, uow):
--> 630         util.preloaded.orm_persistence.save_obj(
    631             self.mapper,
    632             uow.states_for_mapper_hierarchy(self.mapper, False, False),

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py in save_obj(base_mapper, states, uowtransaction, single)
    232         )
    233 
--> 234         _emit_update_statements(
    235             base_mapper,
    236             uowtransaction,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py in _emit_update_statements(base_mapper, uowtransaction, mapper, table, update, bookkeeping)
    996                 )
    997 
--> 998                 c = connection._execute_20(
    999                     statement, multiparams, execution_options=execution_options
   1000                 )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_20(self, statement, parameters, execution_options)
   1581             )
   1582         else:
-> 1583             return meth(self, args_10style, kwargs_10style, execution_options)
   1584 
   1585     def exec_driver_sql(

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/sql/elements.py in _execute_on_connection(self, connection, multiparams, params, execution_options, _force)
    321     ):
    322         if _force or self.supports_execution:
--> 323             return connection._execute_clauseelement(
    324                 self, multiparams, params, execution_options
    325             )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_clauseelement(self, elem, multiparams, params, execution_options)
   1450             linting=self.dialect.compiler_linting | compiler.WARN_LINTING,
   1451         )
-> 1452         ret = self._execute_context(
   1453             dialect,
   1454             dialect.execution_ctx_cls._init_compiled,

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1812 
   1813         except BaseException as e:
-> 1814             self._handle_dbapi_exception(
   1815                 e, statement, parameters, cursor, context
   1816             )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   1993                 util.raise_(newraise, with_traceback=exc_info[2], from_=e)
   1994             elif should_wrap:
-> 1995                 util.raise_(
   1996                     sqlalchemy_exception, with_traceback=exc_info[2], from_=e
   1997                 )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/util/compat.py in raise_(***failed resolving arguments***)
    205 
    206         try:
--> 207             raise exception
    208         finally:
    209             # credit to

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/base.py in _execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1769                             break
   1770                 if not evt_handled:
-> 1771                     self.dialect.do_execute(
   1772                         cursor, statement, parameters, context
   1773                     )

~/Documents/GitHub/disk-objectstore/.tox/py38/lib/python3.8/site-packages/sqlalchemy/engine/default.py in do_execute(self, cursor, statement, parameters, context)
    715 
    716     def do_execute(self, cursor, statement, parameters, context=None):
--> 717         cursor.execute(statement, parameters)
    718 
    719     def do_execute_no_params(self, cursor, statement, context=None):

InterfaceError: (sqlite3.InterfaceError) Error binding parameter 0 - probably unsupported type.
[SQL: UPDATE user SET name=? WHERE user.id = ?]
[parameters: (<object object at 0x7f85a2f51b00>, 1)]
(Background on this error at: https://sqlalche.me/e/14/rvf5)

```